### PR TITLE
Adds special resin walls

### PR DESCRIPTION
**Pull request by Barnet2**
## About The Pull Request
Adds three new limited walls: Bulletproof, Hardy, Fireproof
Bulletproof: Immune to bullets without extremely high AP
Hardy: 1.5x health, same starting health as other walls but higher regen to discourage combat walling
Fireproof: Immune to fire
These use a new ability called "Secrete Special Resin" which has a shared point pool, starting out it is 50 points (up to change)
## Why It's Good For The Game
More xeno defenses is cool.
## Changelog
:cl:
add: Bulletproof, Hardy and Fireproof walls
/:cl:

### CODE REVIEW

**Comment by Barnet2** [comment]
https://github.com/tgstation/TerraGov-Marine-Corps/assets/20828943/7305ce19-545b-45db-88f7-733eb6f746e4
working video i hope

**Comment by Barnet2** [comment]
https://github.com/tgstation/TerraGov-Marine-Corps/assets/20828943/57701312-bb9c-46a8-922d-b3df22dd5423

works I hope

**Comment by Woker2004** [comment]
> specialwalls.mp4 
> works I hope

Barnet, make the reflective wall deflect with like a 75% chance not a 100%, also make the flame wall fire resistant not fire immune

**Comment by Helg2** [comment]
Does it reflect rockets I presume?

**Comment by Barnet2** [comment]
> > specialwalls.mp4
> > works I hope
> 
> Barnet, make the reflective wall deflect with like a 75% chance not a 100%, also make the flame wall fire resistant not fire immune
No unless kurt forces me to
> Does it reflect rockets I presume?
It has a funny interaction with explosives that I am keeping in for the funny, it actually explodes twice. Once when hitting the wall (destroying it) but also a new rocket appears and does what you would expect it to. Double the explosive filling!!!

File: code/modules/mob/living/carbon/xenomorph/abilities.dm [hunk 1]
```diff
@@ -439,6 +439,176 @@
...
+	. = base_wait
+	if(!scaling_wait)
+		return
+	var/mob/living/carbon/xenomorph/X = owner
+
+	var/build_resin_modifier = 1
+	switch(X.selected_resin)
+		if(/obj/alien/resin/sticky)
+			build_resin_modifier = 0.5
+
```
Comment (DeltaFire15) [inline code/modules/mob/living/carbon/xenomorph/abilities.dm:528]: This is not checking the special resin var (aka, if you have sticky selected on your base action, your special walls will build twice as fast).

File: code/modules/mob/living/carbon/xenomorph/abilities.dm [hunk 2]
```diff
@@ -439,6 +439,176 @@
...
+	target_flags = ABILITY_TURF_TARGET
+	ability_cost = 75
+	action_type = ACTION_TOGGLE
+	keybinding_signals = list(
+		KEYBINDING_NORMAL = COMSIG_XENOABILITY_SECRETE_SPECIAL_RESIN,
+	)
+	///Minimum time to build a resin structure
+	var/base_wait = 1 SECONDS
+	///Multiplicator factor to add to the building time, depends on the health of the structure built
+	var/scaling_wait = 1 SECONDS
```
Comment (DeltaFire15) [inline code/modules/mob/living/carbon/xenomorph/abilities.dm:456]: This is scaled with health damage of the owner, not the structure's health.

**Review by DeltaFire15 — COMMENTED** [review]
Isn't reflective a bit too powerful for combat walling? Essentially warlock shield but it is not limited in duration or (ranged) health. Potentially also absurdly powerful for turret wall strategies or wraith shenanegans.
Every builder caste having an on-demand warlock shield seems a bit too good.

(One second to make too, ,aking that even more concerning.

**Comment by Looking4bros** [comment]
Would add counterplay to the funny Xraygun for defensive structures.

**Comment by SzkieletorNutric** [comment]
Great idea, but careless af

**Comment by Zangeiti** [comment]
reflective walls having 100% accuracy when bounced back seems very guh

**Review by Kuro020 — CHANGES_REQUESTED** [review]
image
would probably be more balance like this

**Comment by MiniMeatwad** [comment]
Apparently someone shot a mortar at the wall, and the mortar bounced back and gibbed a marine.

**Comment by AwkwardStereo** [comment]
https://github.com/tgstation/TerraGov-Marine-Corps/assets/35969040/a8ee2679-0ee9-4bbc-aca7-1ccbeb006876

There was a 2nd reflective tile 2 tiles north of it that bounced projectiles between the two at very hihg incridbl sped that left behind a tile that instantly crit anyone (marine and xeno) who walked over it.

Like the old bouncy lasers deathtiles from a while back.

**Comment by zzurge** [comment]
Yeahhhh reflective walls are a bit silly ngl

**Comment by Fawcksy** [comment]
Screenshot_168
this shit was all it took to kill a FOB because sentries will just shoot whatever, get rid of this garbage. Xenoids have enough buffs without punishing marines for doing the one thing marines do good. Swear to god some of yall take one sniper bullet and write a whole new codebase about it

**Comment by ju45he** [comment]
Pre-round lobby talk around the time I am making this comment said something about a laser sniper on the richochet setting crashing the server when having some interaction with a reflective wall. I didn't quite stick around long enough to really confirm or anything but I just wanna letcha all know that.

**Comment by mkvrs** [comment]
golden age of combat walling buffs

**Comment by MiniMeatwad** [comment]
> Swear to god some of yall take one T60 bullet and write a whole new codebase about it

enjoyingthet60-1

**Comment by coldironwarrior** [comment]
> Pre-round lobby talk around the time I am making this comment said something about a laser sniper on the richochet setting crashing the server when having some interaction with a reflective wall. I didn't quite stick around long enough to really confirm or anything but I just wanna letcha all know that.

i like the concept of this PR; i thought Better Mazing would be cool.
but a number of our systems and implementations don't play nicely with this. including the magic death tiles [which stereo and gruch documented above]; what ju45he mentions above [round 25847 ended pre-maturely because of this, round 25848 ran into issues with this as well, but not enough to outright kill the server]. there's some static images but i'm not sure anyone has a video. richocet laser when bouncing creates a new projectile anyways; so when it hits a reflective wall, it creates...a few projectiles...which hit a reflective wall, which...you get the idea. what fawcks pointed out is also of concern to me, this is a cheap way to destroy sentries and take out a fob specifically with little counterplay.

fireproof walls when flamed obviously are fine but allow the weeds on tiles below them to be destroyed, which seems a little off.

File: code/datums/keybinding/xeno.dm [hunk 3]
```diff
@@ -47,6 +47,12 @@
 	keybind_signal = COMSIG_XENOABILITY_SECRETE_RESIN
 	hotkey_keys = list("R")
 
+/datum/keybinding/xeno/secrete_special_resin
+	name = "secrete_special_resin"
+	full name = "Secrete Special Resin"
+	description = "Builds whatever special resin you have selected."
+	keybind_signal = COMSIG_XENOABILITY_SECRETE_SPECIAL_RESIN
+	hotkey_keys = list("ShiftR")
 /datum/keybinding/xeno/recycle
```
Comment (TiviPlus) [inline code/datums/keybinding/xeno.dm:56]:
```suggestion
	hotkey_keys = list("ShiftR")

/datum/keybinding/xeno/recycle
```

File: code/game/turfs/walls/resin.dm [hunk 4]
```diff
@@ -205,4 +205,41 @@
 
 /* Hivelord walls, they start off stronger */
 /turf/closed/wall/resin/regenerating/thick
-	max_integrity = 125
+	max_integrity = 250
+
```
Comment (TiviPlus) [inline code/game/turfs/walls/resin.dm:209]: ???
Undocumented change?

File: code/game/turfs/walls/resin.dm [hunk 5]
```diff
@@ -205,4 +205,41 @@
...
+	X.visible_message(span_xenonotice("\The [X] tears down \the [src]!"), \
+	span_xenonotice("We tear down \the [src]."))
+	playsound(src, "alien_resin_break", 25)
+	take_damage(max_integrity) // Ensure its destroyed
+
+/turf/closed/wall/resin/regenerating/special/bulletproof
+	name = "bulletproof resin wall"
+	desc = "Weird slime solidified into a wall. Looks shiny."
+	soft_armor = list(MELEE = 0, BULLET = 100, LASER = 100, ENERGY = 100, BOMB = 0, BIO = 0, FIRE = 0, ACID = 0) //You aren't damaging this with bullets without alot of AP.
+	color = "#ed99f6"
```
Comment (TiviPlus) [inline code/game/turfs/walls/resin.dm:232]: use color defines

File: code/game/turfs/walls/resin.dm [hunk 6]
```diff
@@ -205,4 +205,41 @@
...
+		return
+	X.visible_message(span_xenonotice("\The [X] starts tearing down \the [src]!"), \
+	span_xenonotice("We start to tear down \the [src]."))
+	if(!do_after(X, 1 SECONDS, NONE, X, BUSY_ICON_GENERIC))
+		return
+	if(!istype(src)) // Prevent jumping to other turfs if do_after completes with the wall already gone
+		return
+	X.do_attack_animation(src, ATTACK_EFFECT_CLAW)
+	X.visible_message(span_xenonotice("\The [X] tears down \the [src]!"), \
+	span_xenonotice("We tear down \the [src]."))
```
Comment (TiviPlus) [inline code/game/turfs/walls/resin.dm:224]: isnt this done on parent

File: code/modules/mob/living/carbon/xenomorph/abilities.dm [hunk 7]
```diff
@@ -439,6 +439,169 @@
...
+	///Minimum time to build a special resin structure
+	var/base_wait = 2 SECONDS
+	///Multiplicator factor to add to the building time, depends on the health of builder.
+	var/scaling_wait = 1 SECONDS
+	///List of buildable structures. Order corresponds with resin_special_images_list.
+	var/list/buildable_special_structures = list(
+		/turf/closed/wall/resin/regenerating/special/bulletproof,
+		/turf/closed/wall/resin/regenerating/special/fireproof,
+		/turf/closed/wall/resin/regenerating/special/hardy
+		)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/abilities.dm:462]:
```suggestion
		/turf/closed/wall/resin/regenerating/special/hardy,
	)
```
suprised regex didnt catch this

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/game/turfs/walls/resin.dm (hunk 4 above)
Comment (Barnet2) [reply code/game/turfs/walls/resin.dm:209]: Oops as fuck, when I used merge conflict I forgot about this. With the resin armor changes I had to reduce integ to 125, 250 was the original and it isn't anymore.

File: code/game/turfs/walls/resin.dm (hunk 6 above)
Comment (Barnet2) [reply code/game/turfs/walls/resin.dm:224]: Parent includes refunding for Quickbuild points, this subtype removes that. A better way could to have a check in parent itself maybe? Tell me what you want w/ that.

File: code/game/turfs/walls/resin.dm [hunk 8]
```diff
@@ -91,9 +94,10 @@
 	if(X.status_flags & INCORPOREAL)
 		return
 	if(CHECK_BITFIELD(SSticker.mode?.flags_round_type, MODE_ALLOW_XENO_QUICKBUILD) && SSresinshaping.active)
-		SSresinshaping.quickbuild_points_by_hive[X.hivenumber]++
-		take_damage(max_integrity) // Ensure its destroyed
-		return
+		if(src.is_normal_resin_wall)
```
Comment (TiviPlus) [inline code/game/turfs/walls/resin.dm:97]: src.var

File: code/modules/mob/living/carbon/xenomorph/abilities.dm [hunk 9]
```diff
@@ -439,6 +439,169 @@
...
+			return
+		if(ERROR_FOG)
+			owner.balloon_alert(owner, span_notice("The fog will prevent the resin from ever taking shape!"))
+			return
+		if(ERROR_CONSTRUCT)
+			return
+		if(TRUE)
+			return
+	if(!line_of_sight(owner, T))
+		to_chat(owner, span_warning("You cannot secrete resin without line of sight!"))
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/abilities.dm:556]: balloon alert?

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm [hunk 10]
```diff
@@ -27,6 +27,7 @@
 	var/tier2_xeno_limit
 	/// Queue of all clients wanting to join xeno side
 	var/list/client/candidates
+	var/special_build_points = 50
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/hive_datum.dm:30]: autodoc

**Review by TiviPlus — CHANGES_REQUESTED** [review]

**Review by TiviPlus — APPROVED** [review]

**Pull request merged**